tobac/utils/general.py

```diff
@@ -553,3 +554,129 @@ def combine_tobac_feats(list_of_feats, preserve_old_feat_nums=None):
+        new_dataset, latitude_name=latitude_name, longitude_name=longitude_name
+    )
+
+    if lat_coord not in features or lon_coord not in features:
+        raise ValueError("Cannot find latitude and/or longitude coordinate")
+
+    lat_vals_new = new_dataset[lat_coord].values
+    lon_vals_new = new_dataset[lon_coord].values
+
+    if len(lat_vals_new.shape) != len(lon_vals_new.shape):
+        raise ValueError(
+            "Cannot work with lat/lon coordinates of unequal dimensionality"
+        )
+
+    # the lat/lons must be a 2D grid, so if they aren't, make them one.
+    if len(lat_vals_new.shape) == 1:
+        lon_vals_new, lat_vals_new = np.meshgrid(lon_vals_new, lat_vals_new)
+
+    # we have to convert to radians because scikit-learn's haversine
+    # requires that the input be in radians.
+    flat_lats = np.deg2rad(lat_vals_new.flatten())
+    flat_lons = np.deg2rad(lon_vals_new.flatten())
+
+    # we have to drop NaN values.
+    either_nan = np.logical_or(np.isnan(flat_lats), np.isnan(flat_lons))
+    # we need to remember where these values are in the array so that we can
+    # appropriately unravel them.
+    loc_arr = np.arange(0, len(flat_lats), 1)
+    loc_arr_trimmed = loc_arr[~either_nan]
+    flat_lats_nona = flat_lats[~either_nan]
+    flat_lons_nona = flat_lons[~either_nan]
+    ll_tree = sklearn.neighbors.BallTree(
+        np.array([flat_lats_nona, flat_lons_nona]).T, metric="haversine"
+    )
+    new_h1 = list()
+    new_h2 = list()
+    # there is almost certainly room for speedup in here.
+    for index in features["index"]:
+        dist, closest_pt = ll_tree.query(
+            [
+                [
+                    np.deg2rad(features["latitude"][index]),
+                    np.deg2rad(features["longitude"][index]),
+                ]
+            ]
+        )
+        unraveled_h1, unraveled_h2 = np.unravel_index(
+            loc_arr_trimmed[closest_pt[0][0]], np.shape(lat_vals_new)
+        )
+
+        new_h1.append(unraveled_h1)
+        new_h2.append(unraveled_h2)
+    ret_features = copy.deepcopy(features)
+    ret_features["hdim_1"] = ("index", new_h1)
+    ret_features["hdim_2"] = ("index", new_h2)
+
+    # find where distances are too large and drop them.
+    new_lat = lat_vals_new[new_h1, new_h2]
+    new_lon = lon_vals_new[new_h1, new_h2]
+    dist_apart = (
+        tb_analysis.haversine(
+            new_lat, new_lon, features[lat_coord], features[lon_coord]
+        )
+        * 1000.0
+    )
+    ret_features = ret_features.where(dist_apart < max_space_away, drop=True)
+
+    # force times to match, where appropriate.
+    if "time" in new_dataset.coords:
+        # this is necessary due to the iris/xarray/pandas weirdness that we have.
+        old_feat_times = ret_features["time"].astype("datetime64[s]")
+        closest_times = np.min(np.abs(old_feat_times - new_dataset["time"]), axis=1)
+        closest_time_locs = np.abs(old_feat_times - new_dataset["time"]).argmin(axis=1)
```

Comment:
This can easily result in memory issues when the total number of features and the total number of timesteps in the new dataset are large. However, iterating over each feature will be also quite ineffective, so we might have to accept that ?

Reply:
Yes, I agree that there are issues on either side. I am inclined to leave it for now, though? It's entirely possible I'll have to optimize this for some upcoming work, though.

Reply:
All right, that is OK for me.